[JSC] Implement SetterThatIgnoresPrototypeProperties() as specified in Iterator helpers proposal

shvaikalesh opened the pull request:
#### a3e7b8f85e8fb9444ed717f3d494052adacddfef
<pre>
[JSC] Implement SetterThatIgnoresPrototypeProperties() as specified in Iterator helpers proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=280705">https://bugs.webkit.org/show_bug.cgi?id=280705</a>
&lt;<a href="https://rdar.apple.com/problem/137078675">rdar://problem/137078675</a>&gt;

Reviewed by Yusuke Suzuki.

Unfortunately, putDirect() is not enough to implement SetterThatIgnoresPrototypeProperties() [1]:
  * if the property is absent, and object is non-extensible, a TypeError should be thrown;
  * if the property is present, its descriptor should not be overwritten (like putDirect() does),
    and the setter, if present, should be invoked.

Also, SetterThatIgnoresPrototypeProperties() should invoke non-ordinary [[GetOwnProperty]], [[Set]],
and [[DefineOwnProperty]] methods because it can be called on any object, including a ProxyObject.

[1]: <a href="https://tc39.es/proposal-iterator-helpers/#sec-SetterThatIgnoresPrototypeProperties">https://tc39.es/proposal-iterator-helpers/#sec-SetterThatIgnoresPrototypeProperties</a> (steps 3-5)

* JSTests/stress/iterator-prototype-constructor-setter.js:
* JSTests/stress/iterator-prototype-to-string-tag-basic-setter.js:
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSC_DEFINE_CUSTOM_SETTER):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::setterThatIgnoresPrototypeProperties):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: <a href="https://commits.webkit.org/284737@main">https://commits.webkit.org/284737@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/adfadb6a773e237ce4f5ec538b2a7bb74a7541b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/70368 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/49774 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/21546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/57574 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/21386 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/21546 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/73434 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/57574 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/57574 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/19907 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/63489 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/57574 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/69615 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/87/builds/14594 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/21386 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/76176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/14636 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/76176 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/23133 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/91398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10767 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/45576 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/343 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/91398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/46650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/47927 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/46392 "EWS skipped this build as PR had skip-ews label when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -1750,6 +1750,8 @@ bool validateAndApplyPropertyDescriptor(JSGlobalObject*, JSObject*, PropertyName
 JS_EXPORT_PRIVATE NEVER_INLINE bool ordinarySetSlow(JSGlobalObject*, JSObject*, PropertyName, JSValue, JSValue receiver, bool shouldThrow);
 JS_EXPORT_PRIVATE NEVER_INLINE bool ordinarySetWithOwnDescriptor(JSGlobalObject*, JSObject*, PropertyName, JSValue, JSValue receiver, PropertyDescriptor&& ownDescriptor, bool shouldThrow);
 
+JS_EXPORT_PRIVATE NEVER_INLINE bool setterThatIgnoresPrototypeProperties(JSGlobalObject*, JSValue thisValue, JSObject* homeObject, PropertyName, JSValue, bool shouldThrow);
```
I think we do not need to use `JS_EXPORT_PRIVATE` since nobody outside of JSC is using this function.

shvaikalesh replied:
Fixed.

Constellation reviewed (approved):
r=me

Commit:
[JSC] Implement SetterThatIgnoresPrototypeProperties() as specified in Iterator helpers proposal
https://bugs.webkit.org/show_bug.cgi?id=280705
<rdar://problem/137078675>

Reviewed by Yusuke Suzuki.

Unfortunately, putDirect() is not enough to implement SetterThatIgnoresPrototypeProperties() [1]:
  * if the property is absent, and object is non-extensible, a TypeError should be thrown;
  * if the property is present, its descriptor should not be overwritten (like putDirect() does),
    and the setter, if present, should be invoked.

Also, SetterThatIgnoresPrototypeProperties() should invoke non-ordinary [[GetOwnProperty]], [[Set]],
and [[DefineOwnProperty]] methods because it can be called on any object, including a ProxyObject.

[1]: https://tc39.es/proposal-iterator-helpers/#sec-SetterThatIgnoresPrototypeProperties (steps 3-5)

* JSTests/stress/iterator-prototype-constructor-setter.js:
* JSTests/stress/iterator-prototype-to-string-tag-basic-setter.js:
* Source/JavaScriptCore/runtime/JSIteratorPrototype.cpp:
(JSC::JSC_DEFINE_CUSTOM_SETTER):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::setterThatIgnoresPrototypeProperties):
* Source/JavaScriptCore/runtime/JSObject.h:

Canonical link: https://commits.webkit.org/284737@main
